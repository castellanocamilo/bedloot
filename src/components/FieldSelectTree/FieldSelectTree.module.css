@import '../../styles/customMediaQueries.css';

.root {
  margin-bottom: 16px;
}

.optionList {
  margin: 0;

  .optionList {
    margin: 3px 0;
  }

  @media (--viewportMedium) {
    .optionList {
      margin: 4px 0;
    }
  }
}

.option {
  padding-left: 12px;
}

.optionBtn {
  composes: h4 from global;
  font-weight: var(--fontWeightMedium);
  font-size: 16px;
  line-height: 24px;
  color: var(--colorGrey700);
  text-align: left;

  /* Layout */
  display: block;
  position: relative;
  width: 100%;
  margin: 0;
  padding: 4px 0 8px 0;

  /* Override button styles */
  outline: none;
  border: none;
  cursor: inherit;

  &:focus,
  &:hover {
    color: var(--colorBlack);
  }

  @media (--viewportMedium) {
    padding: 8px 0 0px 0;
    padding: 0;
  }
}

.optionBtnSelected {
  composes: optionBtn;
  color: var(--marketplaceColor);

  &:focus,
  &:hover {
    color: var(--marketplaceColorDark);
  }
}

.optionBtnSelectedLowest {
  &:focus,
  &:hover {
    color: var(--marketplaceColor);
  }
}

.longWord {
  word-break: break-word;
  hyphens: auto;
}
