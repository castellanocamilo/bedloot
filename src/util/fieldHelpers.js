// This util file is about user and listing fields.
// I.e. These are custom fields to data entities. They are added through the Marketplace Console.
// In the codebase, we also have React Final Form fields, which are wrapper around user inputs.

import { SCHEMA_TYPE_MULTI_ENUM, SCHEMA_TYPE_TEXT } from './types';

const keyMapping = {
  userType: {
    wrapper: 'userTypeConfig',
    limitTo: 'limitToUserTypeIds',
    ids: 'userTypeIds',
  },
  listingType: {
    wrapper: 'listingTypeConfig',
    limitTo: 'limitToListingTypeIds',
    ids: 'listingTypeIds',
  },
  category: {
    wrapper: 'categoryConfig',
    limitTo: 'limitToCategoryIds',
    ids: 'categoryIds',
  },
};

const getEntityTypeRestrictions = (entityTypeKey, config) => {
  const { wrapper, limitTo, ids } = keyMapping[entityTypeKey];
  const isLimited = config && config[wrapper] && config[wrapper][limitTo];
  const limitToIds = config && config[wrapper] && config[wrapper][ids];

  return { isLimited, limitToIds };
};

/**
 * Check if the given listing type is allowed according to the given listing field config.
 *
 * @param {String} entityTypeKey entity type key (e.g. 'listingType', 'userType')
 * @param {String|Array<String>} entityType entity type to be checked (e.g. 'amenities'). Accepts an array too.
 * @param {*} fieldConfig the config of a custom listing field
 * @returns true if listingTypeConfig allows the listingType
 */
export const isFieldFor = (entityTypeKey, entityType, fieldConfig) => {
  const { isLimited, limitToIds } = getEntityTypeRestrictions(entityTypeKey, fieldConfig);

  if (Array.isArray(entityType)) {
    return !isLimited || limitToIds.some(cid => entityType.includes(cid));
  }
  return !isLimited || limitToIds.includes(entityType);
};

export const isFieldForUserType = (userType, fieldConfig) =>
  isFieldFor('userType', userType, fieldConfig);
export const isFieldForListingType = (listingType, fieldConfig) =>
  isFieldFor('listingType', listingType, fieldConfig);
export const isFieldForCategory = (categories, fieldConfig) =>
  isFieldFor('category', categories, fieldConfig);

/**
 * Returns the value of the attribute in extended data.
 * @param {*} data extended data containing the value
 * @param {*} key attribute key in extended data
 * @returns
 */
export const getFieldValue = (data, key) => {
  const value = data?.[key];
  return value != null ? value : null;
};

/**
 * Picks current values for listing categories based on provided public data and configuration.
 * This function validates if the initial values match with the configuration received via assets.
 * If a categoryLevel value doesn't match with the category configuration, it is not passed on to the form.
 *
 * @param {*} data publicData or some other set where category-related nested data is available
 * @param {String} prefix prefix used for storing nested values.
 * @param {Number} level refers to nesting level (starts from 1)
 * @param {Array} categoryLevelOptions array of nested category structure
 * @returns pick valid prefixed properties
 */
export const pickCategoryFields = (data, prefix, level, categoryLevelOptions = []) => {
  const currentCategoryKey = `${prefix}${level}`;
  const currentCategoryValue = data[currentCategoryKey];
  const isCategoryLevelSet = typeof currentCategoryValue !== 'undefined';

  // Validate the value against category options
  const categoryOptionConfig = categoryLevelOptions.find(
    category => category.id === currentCategoryValue
  );
  const isValidCategoryValue = !!categoryOptionConfig;
  const nextLevelOptions = categoryOptionConfig?.subcategories || [];

  // Return category level property if it's found from the data and the value is one of the valid options.
  // Go through all the nested levels.
  return isCategoryLevelSet && isValidCategoryValue
    ? {
        [currentCategoryKey]: currentCategoryValue,
        ...pickCategoryFields(data, prefix, ++level, nextLevelOptions),
      }
    : {};
};

/**
 * Pick props for SectionMultiEnumMaybe and SectionTextMaybe display components.
 *
 * @param {*} publicData entity public data containing the value(s) to be displayed
 * @param {*} metadata entity metadata containing the value(s) to be displayed
 * @param {Array<Object>} fieldConfigs array of custom field configuration objects
 * @param {String} entityTypeKey the name of the key denoting the entity type in publicData.
 * E.g. 'listingType', 'userType', or 'category'
 * @param {function} shouldPickFn an optional function to add extra check before including the field props
 * @returns an object with attributes 'schemaType', 'key', and 'heading', as well as either
 * - 'options' and 'selectedOptions' for SCHEMA_TYPE_MULTI_ENUM
 * - or 'text' for SCHEMA_TYPE_TEXT
 */
export const pickCustomFieldProps = (
  publicData,
  metadata,
  fieldConfigs,
  entityTypeKey,
  shouldPickFn
) => {
  return fieldConfigs?.reduce((pickedElements, config) => {
    const { key, enumOptions, schemaType, scope = 'public', showConfig } = config;
    const { label, unselectedOptions: showUnselectedOptions } = showConfig || {};
    const entityType = publicData && publicData[entityTypeKey];
    const isTargetEntityType = isFieldFor(entityTypeKey, entityType, config);

    const createFilterOptions = options =>
      options.map(o => ({ key: `${o.option}`, label: o.label }));

    const shouldPick = shouldPickFn ? shouldPickFn(config) : true;

    const value =
      scope === 'public'
        ? getFieldValue(publicData, key)
        : scope === 'metadata'
        ? getFieldValue(metadata, key)
        : null;

    return isTargetEntityType && schemaType === SCHEMA_TYPE_MULTI_ENUM && shouldPick
      ? [
          ...pickedElements,
          {
            schemaType,
            key,
            heading: label,
            options: createFilterOptions(enumOptions),
            selectedOptions: value || [],
            showUnselectedOptions: showUnselectedOptions !== false,
          },
        ]
      : isTargetEntityType && !!value && config.schemaType === SCHEMA_TYPE_TEXT && shouldPick
      ? [
          ...pickedElements,
          {
            schemaType,
            key,
            heading: label,
            text: value,
          },
        ]
      : pickedElements;
  }, []);
};
