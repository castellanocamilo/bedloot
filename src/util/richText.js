import React from 'react';
import flow from 'lodash/flow';
import flatMap from 'lodash/flatMap';
import map from 'lodash/map';

import { sanitizeUrl } from './sanitize';

import { ExternalLink } from '../components';
// NOTE: This file imports components/index.js, which may lead to circular dependency

/**
 * Add zero width space (zwsp) around given breakchars (default '/') to make word break possible.
 * E.g. "one/two/three" => ["one", "​/​", "two" "​/​" "three"]
 *
 * @param {String} wordToBreak word to be broken from special character points.
 * @param {String} breakChars string containing possible chars that can be surrounded with zwsp.
 * @return {Array<String>} returns an array containing strings-
 */
export const zwspAroundSpecialCharsSplit = (wordToBreak, breakChars = '/') => {
  if (typeof wordToBreak !== 'string') {
    return wordToBreak;
  }

  const bcArray = breakChars.split('');

  // Escape special regular expression chars
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions
  const escapedBCArray = bcArray.map(c => c.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'));
  const reSplit = new RegExp('([' + escapedBCArray.join('') + '])');

  const zwsp = '​';
  return wordToBreak.split(reSplit).map(w => (bcArray.includes(w) ? `${zwsp}${w}${zwsp}` : w));
};

/**
 * Layouts are not fixed sizes - So, long words in text make flexboxed items to grow too big.
 * This wraps long words with span and adds given class to it
 *
 * @param {String} word to be wrapped if requirement (longWordMinLength) is met
 * @param {Number} key span needs a key in React/JSX
 * @param {Number} longWordMinLength minimum length when word is considered long
 * @param {String} longWordClass class to be added to spans
 * @return {node} returns a string or component
 */
export const wrapLongWord = (word, key, options = {}) => {
  const { longWordMinLength, longWordClass } = options;
  if (typeof word !== 'string' || !(longWordMinLength && longWordClass)) {
    return word;
  }

  const isShortWord = word.length <= longWordMinLength;
  return isShortWord ? (
    word
  ) : (
    <span key={key} className={longWordClass}>
      {word}
    </span>
  );
};

/**
 * Find links from words and surround them with <ExternalLink> component
 *
 * @param {String} word to be linkified if requirement (link) is met
 * @param {Number} key span needs a key in React/JSX
 * @param {Object} options than can contain keys: linkify, linkClass.
 * @return {Array<node>} returns a array containing ExternalLink component or strings
 */
export const linkifyOrWrapLinkSplit = (word, key, options = {}) => {
  if (typeof word !== 'string') {
    return word;
  }
  const { linkify, linkClass } = options;

  // TODO This can't handle links that contain parenthesis:
  // '(http://example.org/path_(etc))'
  // Currently extracts:
  // '(<a href=\"http://example.org/path_\" ...>http://example.org/path_</a>(etc))'
  //
  // We need to
  // 1) track whether token before link contains parenthesis as a last character
  //    before link token ("word.split(urlRegex)[linkIndex - 1]") and
  // 2) add enough characters to the end of link-token from the next token
  //    after link ("word.split(urlRegex)[linkIndex + 1]")

  // urlRegex modified from examples in
  // https://stackoverflow.com/questions/1500260/detect-urls-in-text-with-javascript

  // eslint-disable-next-line no-useless-escape
  const urlRegex = /(\bhttps?:\/\/[-A-Z0-9+&@#\/%?=~_|!:,.;]*[-A-Z0-9+&@#\/%=~_|])/gi;
  if (word.match(urlRegex)) {
    // Split strings like "(http://www.example.com)" to ["(","http://www.example.com",")"]
    return word.split(urlRegex).map(w => {
      const isEmptyString = !w.match(urlRegex);
      const sanitizedURL = !isEmptyString && linkify ? sanitizeUrl(w) : w;
      return isEmptyString ? (
        w
      ) : linkify ? (
        <ExternalLink key={key} href={sanitizedURL} className={linkClass}>
          {sanitizedURL}
        </ExternalLink>
      ) : linkClass ? (
        <span key={key} className={linkClass}>
          {w}
        </span>
      ) : (
        w
      );
    });
  } else {
    return word;
  }
};

/**
 * Scan text to fill in wrappers for long words and add links.
 * Wrap long words: options should contain longWordMinLength & longWordClass
 * Linkify found links: options should contain "linkify: true" (linkClass is optional)
 *
 * Note: this autolinks only strings that start with 'http'. In addition, links are assumed
 *       to lead outside of the app. In-app linking is not supported atm.
 *
 * @param {String} text check text content
 * @param {Object} options { longWordMinLength, longWordClass, linkify = false, linkClass }
 * @return {Array<node>} returns a child array containing strings and inline elements
 */
export const richText = (text, options) => {
  if (typeof text !== 'string') {
    return text;
  }

  // longWordMinLength & longWordClass are needed for long words to be spanned
  // linkify = true is needed for links to be linkified (linkClass is optional)
  const { longWordMinLength, longWordClass, linkify = false, linkClass, breakChars } = options;
  const linkOrLongWordClass = linkClass ? linkClass : longWordClass;
  const nonWhiteSpaceSequence = /([^\s]+)/gi;
  const breakCharsConfig = breakChars != null ? breakChars : '/,';

  return text.split(nonWhiteSpaceSequence).reduce((acc, nextChild, i) => {
    const parts = flow([
      v =>
        flatMap(v, w => linkifyOrWrapLinkSplit(w, i, { linkify, linkClass: linkOrLongWordClass })),
      v => flatMap(v, w => zwspAroundSpecialCharsSplit(w, breakCharsConfig)),
      v => map(v, (w, j) => wrapLongWord(w, `${i}${j}`, { longWordMinLength, longWordClass })),
    ])([nextChild]);
    return acc.concat(parts);
  }, []);
};
