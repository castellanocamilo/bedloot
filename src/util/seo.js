const ensureOpenGraphLocale = locale => {
  switch (locale) {
    case 'en':
      return 'en_US';
    default:
      return locale;
  }
};

/**
 * These will be used with Helmet <meta {...openGraphMetaProps} />
 */
export const openGraphMetaProps = data => {
  const {
    openGraphType,
    socialSharingTitle,
    socialSharingDescription,
    published,
    updated,
    url,
    locale,
    facebookImages,
    facebookAppId,
    marketplaceName,
  } = data;

  if (!(socialSharingTitle && socialSharingDescription && openGraphType && url && facebookImages)) {
    /* eslint-disable no-console */
    if (console && console.warn) {
      console.warn(
        `Can't create Open Graph meta tags:
        socialSharingTitle, socialSharingDescription, openGraphType, url, and facebookImages are needed.`
      );
    }
    /* eslint-enable no-console */
    return [];
  }

  const openGraphMeta = [
    { property: 'og:description', content: socialSharingDescription },
    { property: 'og:title', content: socialSharingTitle },
    { property: 'og:type', content: openGraphType },
    { property: 'og:url', content: url },
    { property: 'og:locale', content: ensureOpenGraphLocale(locale) },
  ];

  if (facebookImages && facebookImages.length > 0) {
    facebookImages.forEach(i => {
      openGraphMeta.push({
        property: 'og:image',
        content: i.url,
      });

      if (i.width && i.height) {
        openGraphMeta.push({ property: 'og:image:width', content: i.width });
        openGraphMeta.push({ property: 'og:image:height', content: i.height });
      }
    });
  }

  if (marketplaceName) {
    openGraphMeta.push({ property: 'og:site_name', content: marketplaceName });
  }

  if (facebookAppId) {
    openGraphMeta.push({ property: 'fb:app_id', content: facebookAppId });
  }

  if (published) {
    openGraphMeta.push({ property: 'article:published_time', content: published });
  }

  if (updated) {
    openGraphMeta.push({ property: 'article:modified_time', content: updated });
  }

  return openGraphMeta;
};

/**
 * These will be used with Helmet <meta {...twitterMetaProps} />
 */
export const twitterMetaProps = data => {
  const {
    socialSharingTitle,
    socialSharingDescription,
    twitterHandle,
    twitterImages,
    url,
    marketplaceRootURL,
    siteTwitterHandle,
  } = data;

  if (!(socialSharingTitle && socialSharingDescription && url)) {
    /* eslint-disable no-console */
    if (console && console.warn) {
      console.warn(
        `Can't create twitter card meta tags:
        socialSharingTitle, socialSharingDescription, and url are needed.`
      );
    }
    /* eslint-enable no-console */
    return [];
  }

  const twitterMeta = [
    { name: 'twitter:card', content: 'summary_large_image' },
    { name: 'twitter:title', content: socialSharingTitle },
    { name: 'twitter:description', content: socialSharingDescription },
    { name: 'twitter:url', content: url },
  ];

  if (siteTwitterHandle) {
    twitterMeta.push({ name: 'twitter:site', content: siteTwitterHandle });
  }

  if (twitterImages && twitterImages.length > 0) {
    twitterImages.forEach(i => {
      twitterMeta.push({
        name: 'twitter:image',
        content: i.url,
      });
    });
  }

  if (twitterHandle) {
    // TODO: If we want to connect providers twitter account on ListingPage
    // we needs to get this info among listing data (API support needed)
    twitterMeta.push({ name: 'twitter:creator', content: twitterHandle });
  }

  if (marketplaceRootURL) {
    twitterMeta.push({ name: 'twitter:domain', content: marketplaceRootURL });
  }

  return twitterMeta;
};

/**
 * These will be used with Helmet <meta {...metaTagProps} />
 * Creates data for Open Graph and Twitter meta tags.
 */
export const metaTagProps = (tagData, config) => {
  const {
    marketplaceRootURL,
    facebookAppId,
    marketplaceName,
    siteTwitterHandle,
    googleSearchConsole,
  } = config;

  const author = tagData.author || marketplaceName;
  const googleSiteVerification = googleSearchConsole?.googleSiteVerification;
  const googleSiteVerificationMaybe = googleSiteVerification
    ? [{ name: 'google-site-verification', content: googleSiteVerification }]
    : [];
  const defaultMeta = [
    { name: 'description', content: tagData.description },
    { name: 'author', content: author },
    ...googleSiteVerificationMaybe,
  ];

  const openGraphMeta = openGraphMetaProps({
    ...tagData,
    facebookAppId,
    marketplaceName,
  });

  const twitterMeta = twitterMetaProps({
    ...tagData,
    marketplaceRootURL,
    siteTwitterHandle,
  });

  return [...defaultMeta, ...openGraphMeta, ...twitterMeta];
};
