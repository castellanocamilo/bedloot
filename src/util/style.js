/**
 * Construct a list of custom CSS Properties from branding config.
 *
 * @param {Object} brandingConfig branding configuration
 * @returns Object literal containing custom CSS Properties (e.g. ['--marketplaceColor']: #aaff00).
 */
export const getCustomCSSPropertiesFromConfig = brandingConfig => {
  const marketplaceColorsMaybe = brandingConfig.marketplaceColor
    ? {
        ['--marketplaceColor']: brandingConfig.marketplaceColor,
        ['--marketplaceColorDark']: brandingConfig.marketplaceColorDark,
        ['--marketplaceColorLight']: brandingConfig.marketplaceColorLight,
      }
    : {};
  const primaryButtonColorsMaybe = brandingConfig.colorPrimaryButton
    ? {
        ['--colorPrimaryButton']: brandingConfig.colorPrimaryButton,
        ['--colorPrimaryButtonDark']: brandingConfig.colorPrimaryButtonDark,
        ['--colorPrimaryButtonLight']: brandingConfig.colorPrimaryButtonLight,
      }
    : {};

  return { ...marketplaceColorsMaybe, ...primaryButtonColorsMaybe };
};

/**
 * This includes Custom CSS Properties, which are defined in hosted asset: branding.json
 *
 * @param {Object} brandingConfig branding configuration
 * @param {Node} element DOM element, which gets these CSS vars included.
 */
export const includeCSSProperties = (brandingConfig, element) => {
  Object.entries(getCustomCSSPropertiesFromConfig(brandingConfig)).forEach(customCSSProperty => {
    const [key, value] = customCSSProperty;
    element.style.setProperty(key, value);
  });
};
