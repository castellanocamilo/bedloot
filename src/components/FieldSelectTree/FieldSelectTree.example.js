/* eslint-disable no-console */
import React from 'react';
import { Form as FinalForm, FormSpy } from 'react-final-form';

import { pickInitialValuesForFieldSelectTree } from '../../util/search';
import * as validators from '../../util/validators';

import { Button } from '../../components';

import FieldSelectTree from './FieldSelectTree';

const config = {
  options: [
    {
      option: 'dogs',
      label: 'Dogs',
      suboptions: [
        {
          option: 'labradors',
          label: 'Labradors',
        },
        {
          option: 'poodles',
          label: 'Poodles',
        },
      ],
    },
    {
      option: 'cats',
      label: 'Cats',
      suboptions: [
        {
          option: 'burmese',
          label: 'Burmese cats',
        },
        {
          option: 'egyptian-mau',
          label: 'Egyptian mau',
        },
      ],
    },
    {
      option: 'fish',
      label: 'Fish',
      suboptions: [
        {
          option: 'freshwater',
          label: 'Freshwater',
          suboptions: [
            {
              option: 'grayling',
              label: 'Grayling',
            },
            {
              option: 'arctic-char',
              label: 'Arctic char',
            },
            {
              option: 'pike',
              label: 'Pike',
            },
          ],
        },
        {
          option: 'saltwater',
          label: 'Saltwater',
          suboptions: [
            {
              option: 'tuna',
              label: 'Tuna',
            },
            {
              option: 'cod',
              label: 'Cod',
            },
          ],
        },
        {
          option: 'euryhaline',
          label: 'Euryhaline fish',
          suboptions: [
            {
              option: 'salmon',
              label: 'Salmon',
            },
            {
              option: 'trout',
              label: 'Trout',
            },
            {
              option: 'european-eel',
              label: 'European eel',
            },
          ],
        },
      ],
    },
    {
      option: 'birds',
      label: 'Birds',
      suboptions: [
        {
          option: 'parrot',
          label: 'Parrot',
        },
        {
          option: 'macaw',
          label: 'Macaw',
        },
      ],
    },
    {
      option: 'reptiles',
      label: 'Reptiles',
      suboptions: [
        {
          option: 'lizards',
          label: 'Lizards',
        },
        {
          option: 'snakes',
          label: 'Snakes',
        },
      ],
    },
  ],
};

const FormComponent = props => (
  <FinalForm
    {...props}
    render={formRenderProps => {
      const { handleSubmit, onChange, invalid, pristine, submitting } = formRenderProps;
      const submitDisabled = submitting; //invalid || pristine || submitting;
      const required = validators.requiredSelectTreeOption('This field is required');
      return (
        <form
          onSubmit={e => {
            e.preventDefault();
            handleSubmit(e);
          }}
        >
          <FormSpy onChange={onChange} />
          <FieldSelectTree
            label="Nested options"
            name="nestedLevel"
            options={config.options}
            validate={required}
          />
          <Button style={{ marginTop: 24 }} type="submit" disabled={submitDisabled}>
            Submit
          </Button>
        </form>
      );
    }}
  />
);

const initialData = {}; //{ nestedLevel1: 'women', nestedLevel2: 'jackets', foo: 'bar' };

export const SelectNested = {
  component: FormComponent,
  props: {
    initialValues: {
      nestedLevel: pickInitialValuesForFieldSelectTree('nestedLevel', initialData),
    },
    onChange: formState => {
      const nestedLevel = formState.values?.nestedLevel;
      if (nestedLevel && Object.values(nestedLevel)?.length > 0) {
        console.log('field value changed to:', formState.values.nestedLevel);
      }
    },
    onSubmit: values => {
      console.log('submit values:', values);
      return false;
    },
  },
  group: 'inputs',
};
