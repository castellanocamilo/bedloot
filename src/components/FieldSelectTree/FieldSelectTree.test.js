import React from 'react';
import '@testing-library/jest-dom';
import { Form as FinalForm } from 'react-final-form';

import { renderWithProviders as render, testingLibrary } from '../../util/testHelpers';
import * as validators from '../../util/validators';

import FieldSelectTree from './FieldSelectTree';

const { screen, userEvent } = testingLibrary;

const noop = () => {};

const config = {
  options: [
    {
      option: 'dogs',
      label: 'Dogs',
      suboptions: [
        {
          option: 'labradors',
          label: 'Labradors',
        },
        {
          option: 'poodles',
          label: 'Poodles',
        },
      ],
    },
    {
      option: 'cats',
      label: 'Cats',
      suboptions: [
        {
          option: 'burmese',
          label: 'Burmese cats',
        },
        {
          option: 'egyptian-mau',
          label: 'Egyptian mau',
        },
      ],
    },
    {
      option: 'fish',
      label: 'Fish',
      suboptions: [
        {
          option: 'freshwater',
          label: 'Freshwater',
          suboptions: [
            {
              option: 'greyling',
              label: 'Greyling',
            },
            {
              option: 'arctic-char',
              label: 'Arctic char',
            },
            {
              option: 'pike',
              label: 'Pike',
            },
          ],
        },
        {
          option: 'saltwater',
          label: 'Saltwater',
          suboptions: [
            {
              option: 'tuna',
              label: 'Tuna',
            },
            {
              option: 'cod',
              label: 'Cod',
            },
          ],
        },
      ],
    },
    {
      option: 'birds',
      label: 'Birds',
      suboptions: [
        {
          option: 'parrot',
          label: 'Parrot',
        },
        {
          option: 'macaw',
          label: 'Macaw',
        },
      ],
    },
    {
      option: 'reptiles',
      label: 'Reptiles',
      suboptions: [
        {
          option: 'lizards',
          label: 'Lizards',
        },
        {
          option: 'snakes',
          label: 'Snakes',
        },
      ],
    },
  ],
};

const FormComponent = props => (
  <FinalForm
    {...props}
    render={formRenderProps => {
      const { handleSubmit, invalid, pristine, submitting } = formRenderProps;
      const submitDisabled = invalid || pristine || submitting;
      const required = validators.requiredSelectTreeOption('This field is required');
      return (
        <form
          onSubmit={e => {
            e.preventDefault();
            handleSubmit(e);
          }}
        >
          <FieldSelectTree
            label="Nested options"
            name="nestedLevel"
            options={config.options}
            validate={required}
          />
          <button style={{ marginTop: 24 }} type="submit" disabled={submitDisabled}>
            Submit
          </button>
        </form>
      );
    }}
  />
);

const initialData = {}; //{ nestedLevel1: 'women', nestedLevel2: 'jackets', foo: 'bar' };

describe('FieldSelectTree', () => {
  it('matches snapshot', () => {
    const tree = render(<FormComponent onSubmit={noop} initialValues={initialData} />);
    expect(tree.asFragment().firstChild).toMatchSnapshot();
  });

  it('enables submit', () => {
    render(<FormComponent onSubmit={noop} />);

    expect(screen.getByRole('button', { name: 'Submit' })).toBeDisabled();
    expect(screen.getByRole('button', { name: 'Dogs' })).not.toBeDisabled();

    // Select "Fish" and see more options
    expect(screen.queryByRole('button', { name: 'Freshwater' })).not.toBeInTheDocument();
    userEvent.click(screen.getByRole('button', { name: 'Fish' }));
    expect(screen.getByRole('button', { name: 'Freshwater' })).toBeInTheDocument();

    // Select "Freshwater" and see more options
    expect(screen.queryByRole('button', { name: 'Pike' })).not.toBeInTheDocument();
    userEvent.click(screen.getByRole('button', { name: 'Freshwater' }));
    expect(screen.getByRole('button', { name: 'Pike' })).toBeInTheDocument();
  });
});
