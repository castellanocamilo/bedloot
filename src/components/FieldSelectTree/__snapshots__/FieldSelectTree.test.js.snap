// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldSelectTree matches snapshot 1`] = `
<form>
  <div
    class="root"
  >
    <label>
      Nested options
    </label>
    <ul
      class="optionList"
    >
      <li
        class="option"
        style="padding-left: 12px; cursor: pointer;"
      >
        <button
          class="optionBtn"
        >
          Dogs
        </button>
      </li>
      <li
        class="option"
        style="padding-left: 12px; cursor: pointer;"
      >
        <button
          class="optionBtn"
        >
          Cats
        </button>
      </li>
      <li
        class="option"
        style="padding-left: 12px; cursor: pointer;"
      >
        <button
          class="optionBtn"
        >
          Fish
        </button>
      </li>
      <li
        class="option"
        style="padding-left: 12px; cursor: pointer;"
      >
        <button
          class="optionBtn"
        >
          Birds
        </button>
      </li>
      <li
        class="option"
        style="padding-left: 12px; cursor: pointer;"
      >
        <button
          class="optionBtn"
        >
          Reptiles
        </button>
      </li>
    </ul>
  </div>
  <button
    disabled=""
    style="margin-top: 24px;"
    type="submit"
  >
    Submit
  </button>
</form>
`;
